samples/core/multi-device/main.cpp

```diff
@@ -0,0 +1,501 @@
+                }
+            }
+        }
+
+        // Declare output matrix. Output will not be padded.
+        std::vector<cl_float> h_output_grid(output_size, 0);
+
+        // Initialize convolution mask.
+        std::vector<cl_float> h_mask(mask_size);
+        if (diag_opts.verbose)
+        {
+            std::cout << "done.  \nGenerating " << mask_size
+                      << " random numbers for convolution mask...";
+            std::cout.flush();
+        }
+        cl::sdk::fill_with_random(prng, h_mask);
+
+        // Create device buffers, from which we will create the subbuffers for
+        // the subdevices.
+        const cl_uint grid_midpoint = y_dim / 2;
+        const cl_uint pad_grid_midpoint = pad_y_dim / 2;
+
+        if (diag_opts.verbose)
+        {
+            std::cout << "done.\nInitializing device-side storage...";
+            std::cout.flush();
+        }
+
+        cl::Buffer dev_input_grid(context,
+                                  CL_MEM_READ_ONLY | CL_MEM_COPY_HOST_PTR
+                                      | CL_MEM_HOST_NO_ACCESS,
+                                  input_bytes, h_input_grid.data());
+        cl::Buffer dev_output_grid(context,
+                                   CL_MEM_WRITE_ONLY | CL_MEM_ALLOC_HOST_PTR
+                                       | CL_MEM_HOST_READ_ONLY,
+                                   output_bytes);
+        cl::Buffer dev_mask(context,
+                            CL_MEM_READ_ONLY | CL_MEM_COPY_HOST_PTR
+                                | CL_MEM_HOST_NO_ACCESS,
+                            mask_bytes, h_mask.data());
+
+        if (diag_opts.verbose)
+        {
+            std::cout << "done.\nSetting up sub-devices...";
+            std::cout.flush();
+        }
+
+        // Set up subdevices for kernel execution.
+        const size_t half_input_bytes =
+            sizeof(cl_float) * pad_x_dim * (pad_grid_midpoint + 1);
+        const size_t input_offset =
+            sizeof(cl_float) * pad_x_dim * (pad_grid_midpoint - 1);
+        const size_t half_output_bytes =
+            sizeof(cl_float) * x_dim * grid_midpoint;
+
+        std::vector<
+            cl::KernelFunctor<cl::Buffer, cl::Buffer, cl::Buffer, cl_uint2>>
+            convolutions{};
+        std::vector<cl::CommandQueue> sub_queues{};
+        std::vector<cl::Buffer> sub_input_grids{}, sub_output_grids{};
+
+        for (size_t i = 0; i < subdevices.size(); ++i)
+        {
+            auto subdevice = subdevices[i];
+
+            if (diag_opts.verbose)
+            {
+                std::cout
+                    << "\n  Creating kernel and command queue of sub-device "
+                    << i << "...";
+                std::cout.flush();
+            }
+
+            auto convolution =
+                cl::KernelFunctor<cl::Buffer, cl::Buffer, cl::Buffer, cl_uint2>(
+                    program, "convolution_3x3")
+                    .getKernel();
```

Comment:
I think this can be simplified.  Why not:

```c++
auto convolution = cl::Kernel(program, "convolution_3x3");
```

Reply:
This assignment seems to be leftover from a previous logic implemented. I changed it to the simpler version suggested.